Frontend changes to allow adding, modifying, and deleting users

jakekapellen opened the pull request:
This is the frontend portion of the changes. Here's what's new:
- Administrators can add, modify, and delete users. They can edit their own details as well, but they cannot delete themselves.
- During initial setup, you can change the administrator's username, password, e-mail, etc.
- Users can change their own passwords from the Settings screen.
- A variety of small fixes to reduce the chattiness in the console (mostly benign serialization errors and 400/404 issues). There is a fair amount of messiness in parts of the code which I've helped to clean up here and there.

See Librephotos/librephotos#581 for backend pull request.

Fixes Librephotos/librephotos#260 :+1:

jakekapellen committed:
Initial commit before rebranch

jakekapellen committed:
Install mantine modals for confirm popup

jakekapellen committed:
Cleanup and bug fixes for adding/updating users

jakekapellen committed:
Merge branch 'dev' of https://github.com/LibrePhotos/librephotos-frontend into Librephotos-dev

jakekapellen committed:
Merge branch 'Librephotos-dev' into dev

jakekapellen committed:
Minor tweaks

jakekapellen committed:
Refactor password entry component

jakekapellen committed:
Allow users to update their own password

jakekapellen committed:
Misc cleanup to reduce console errors

jakekapellen committed:
Mark new strings for I18N

jakekapellen committed:
Minor cleanup

jakekapellen committed:
Update initial setup to use new modal

jakekapellen committed:
Minor bug fixes

jakekapellen committed:
Minor validation fixes

derneuere commented on src/components/settings/PasswordEntry.tsx:
```diff
@@ -0,0 +1,100 @@
+        ) : (
+          <Text>
+            {t("settings.password.titlechangepassword")}
+            <Button
```
Buttons without Text and just an icon look better as an ActionIcon: https://mantine.dev/core/action-icon/

derneuere replied:
I think you used it also in other places like Admin Area. Would be awesome if you can change it :)

jakekapellen replied:
Good catch - I didn't realize this was a thing... I've converted them to ActionIcons.

derneuere commented:
There is a key missing, when deleting a user:
![error1](https://user-images.githubusercontent.com/6658807/182345547-705c30b5-1e69-41f9-8eea-45305a22361a.png)

derneuere commented:
I think the "Changing the password" component would be better on the profile page, because I would expect it there instead of the settings page.

derneuere commented:
While I like the new modal, it isn't ideal from a user experience when using the setup for the first time.

When you start LibrePhotos for the first time, you get this first time setup:
![step 1](https://user-images.githubusercontent.com/6658807/182347295-cbc94821-03fe-4f48-837e-d20d1fc08d46.png)

If you log in after you created your account, you will see the setup button, with this dialog. It asks the same questions you just answered a couple of seconds ago. It's also not clear, that you have to set up the path:
![step 2](https://user-images.githubusercontent.com/6658807/182347567-837fbd91-224a-4b6d-98b7-afb6a1e38466.png)

It would be better to only ask for the missing information instead.

To improve the first time setup, we could add the path to the first time setup page. However, we would still need the dialog for the demo instance, because we have as a standard demo user, but he doesn't have a path.

jakekapellen committed:
Use ActionIcons instead of buttons, minor fixes

jakekapellen committed:
Move password entry to Profile, I18N stuff

jakekapellen commented:
> There is a key missing, when deleting a user: ![error1](https://user-images.githubusercontent.com/6658807/182345547-705c30b5-1e69-41f9-8eea-45305a22361a.png)

I've fixed this now.

jakekapellen commented:
> I think the "Changing the password" component would be better on the profile page, because I would expect it there instead of the settings page.

This is fixed now.

jakekapellen committed:
Fix missing toast strings

jakekapellen commented:
> There is a key missing, when deleting a user: ![error1](https://user-images.githubusercontent.com/6658807/182345547-705c30b5-1e69-41f9-8eea-45305a22361a.png)

This is fixed now

jakekapellen commented:
> While I like the new modal, it isn't ideal from a user experience when using the setup for the first time.
> 
> When you start LibrePhotos for the first time, you get this first time setup: ![step 1](https://user-images.githubusercontent.com/6658807/182347295-cbc94821-03fe-4f48-837e-d20d1fc08d46.png)
> 
> If you log in after you created your account, you will see the setup button, with this dialog. It asks the same questions you just answered a couple of seconds ago. It's also not clear, that you have to set up the path: ![step 2](https://user-images.githubusercontent.com/6658807/182347567-837fbd91-224a-4b6d-98b7-afb6a1e38466.png)
> 
> It would be better to only ask for the missing information instead.
> 
> To improve the first time setup, we could add the path to the first time setup page. However, we would still need the dialog for the demo instance, because we have as a standard demo user, but he doesn't have a path.

The behavior is already somewhat weird in that you go through the first time setup screen, then get a message saying things aren't set up. The first user created is always the admin, who doesn't necessarily need a path. Should users be able to select their own path, or is that something that only the admin should be able to do? Previously, it was an admin-only privilege, but I could relax that.

My thought for the first-time setup is to modify the button to say "Edit User" and to clarify that the administrator needs to edit their user to specify the directory. I can add more validation to the Edit User popup to require this field in this mode.

jakekapellen commented:
Sorry for all the comments (it seemed easier to push individual fixes on separate commits and reply one-by-one). I've addressed everything except for a suggestion on how to not overly complicate the first-time setup while still making it clear what the administrator needs to do if they want to begin using it.

jakekapellen committed:
Tweak initial setup behavior

jakekapellen commented:
I've implemented the suggestion I had for the initial setup step. It is very simple and simply hooks into the existing Edit User modal, making the directory required before leaving in this case.

polaroidkidd commented on .vscode/settings.json:
```diff
@@ -0,0 +1,8 @@
+{
```
are we committing IDE specific settings?

jakekapellen replied:
Shoot. That was supposed to be ignored. No, this should get removed. Done.

polaroidkidd commented on package.json:
```diff
@@ -40,6 +40,7 @@
     "@mantine/dates": "4.2.12",
     "@mantine/form": "4.2.12",
     "@mantine/hooks": "4.2.12",
+    "@mantine/modals": "^5.0.2",
```
are there no interop issues between @mantine/modals v5+ and the other mantine dependencies?

derneuere replied:
I don't think so, most of the breaking changes happened within the core package. I also checked out the code, and it worked fine.

jakekapellen replied:
I originally tried simply upgrading Mantine to 5.0 across the board, which ended up breaking a LOT of stuff. I fixed it so it would compile and run again, but other unrelated things broke in strange ways, so the juice wasn't worth the squeeze.

I've tested pretty thoroughly, and things seem to work okay. If you feel strongly that I should revert "modals" to 4.2.12, I can, but I'll have to redo the confirm deletion popup (not too hard, thankfully).

derneuere replied:
I think it's fine for now, because we want to migrate to Mantine v5 at some point anyway. I tracked the current progress here: https://github.com/LibrePhotos/librephotos/issues/583

polaroidkidd commented on src/App.tsx:
```diff
@@ -90,64 +91,66 @@ export function App() {
+                  <Route element={<ProtectedRoutes />}>
+                    <Route index element={<TimestampPhotos />} />
+                    <Route path="things" element={<AlbumThing />} />
+                    <Route path="recent" element={<RecentlyAddedPhotos />} />
+                    <Route path="favorites" element={<FavoritePhotos />} />
+                    <Route path="deleted" element={<DeletedPhotos />} />
+                    <Route path="hidden" element={<HiddenPhotos />} />
+                    <Route path="notimestamp" element={<NoTimestampPhotosView />} />
+                    <Route path="useralbums" element={<AlbumUser />} />
+                    <Route path="places" element={<AlbumPlace />} />
+                    <Route path="people" element={<AlbumPeople />} />
+                    <Route path="events" element={<AlbumAuto />} />
+                    <Route path="statistics" element={<Statistics />} />
+                    <Route path="settings" element={<Settings />} />
+                    <Route path="profile" element={<Profile />} />
+                    <Route path="library" element={<Library />} />
+                    <Route path="faces" element={<FaceDashboard />} />
+                    <Route path="search" element={<SearchView />} />
+                    <Route path="person/:albumID" element={<AlbumPersonGallery />} />
+                    <Route path="place/:albumID" element={<AlbumPlaceGallery />} />
+                    <Route path="thing/:albumID" element={<AlbumThingGallery />} />
+                    <Route path="event/:albumID" element={<AlbumAutoGalleryView />} />
+                    <Route path="explorer" element={<Explorer />} />
+                    <Route path="albumviewer" element={<AlbumViewer />} />
+                    <Route path="useralbum/:albumID" element={<AlbumUserGallery />} />
+                    <Route path="shared/tome/:which" element={<SharedToMe />} />
+                    <Route path="shared/fromme/:which" element={<SharedFromMe />} />
+                    <Route path="admin" element={<AdminPage />} />
+                    <Route path="map" element={<PhotoMap />} />
+                    <Route path="placetree" element={<LocationTree />} />
+                    <Route path="wordclouds" element={<WordClouds />} />
+                    <Route path="timeline" element={<Timeline />} />
+                    <Route path="socialgraph" element={<Graph />} />
+                    <Route path="facescatter" element={<FaceScatter />} />
+                    <Route path="countstats" element={<CountStats />} />
+                  </Route>
+                </Routes>
+              </AppShell>
+            </NotificationsProvider>
+          </ModalsProvider>
```
is this just a reformatting or did something change here?

jakekapellen replied:
I added "ModalsProvider." This is a nice feature that can centralize modals handling. For now, it is needed to properly site the "confirm you want to delete a user" modal. Other than that, these lines are just indentation changes.

derneuere replied:
Looks good to me! :+1:

polaroidkidd commented on src/actions/utilActions.js:
```diff
@@ -172,20 +172,35 @@ export function updateAvatar(user, form_data) {
   };
 }
 
+export function deleteUser(user, dispatch) {
+  const username = user.username;
+  Server.delete(`delete/user/${user.id}/`, user)
+    .then(response => {
+      dispatch(fetchUserList());
+      showNotification({
+        message: i18n.t("toasts.deleteuser", { username: username }),
+        title: i18n.t("toasts.deleteusertitle"),
+        color: "teal",
+      });
+    })
+    .catch(error => {
+      console.error(error);
+      dispatch({ type: "DELETE_USER_REJECTED", payload: error });
+    });
+}
+
```
I see you've also put this in the RTK api deffinition. That also generats functions for non-react-functions use. Do you still need this then?

polaroidkidd commented on src/api_client/api.ts:
```diff
@@ -84,6 +85,14 @@ export const api = createApi({
+        body: body,
+        url: "/delete/user/",
+      }),
+      transformResponse: response => UserSchema.parse(response),
```
You might need a try-catch here since the UserSchema.parse function can fail

polaroidkidd commented on src/components/modals/ModalUserEdit.tsx:
```diff
@@ -0,0 +1,363 @@
+            value={userName}
+            error={userNameError}
+            onChange={event => {
+              userToEdit.username = event.currentTarget.value;
+              setUserName(userToEdit.username);
+              validateUsername(event.currentTarget.value);
```
inline functions like these get recreated on every render. In this case every time the user types a single letter.  This also applies to the functions below. Would making use of the `useCallback` hook make sense here?

polaroidkidd commented on src/components/modals/ModalUserEdit.tsx:
```diff
@@ -0,0 +1,363 @@
+
+      <Title order={5}>{t("modalscandirectoryedit.header")} </Title>
+      <Text size="sm" color="dimmed">
+        {t("modalscandirectoryedit.explanation1")} &quot;{userToEdit ? userToEdit.username : "..."}&quot;{" "}
```
a HTML entity exists for the `...`. It's `&hellip;`

derneuere commented:
Looks good to me! Awesome pull request. I think polaroidkidd found a couple of things though :sweat_smile: But we can merge it after that.

The paths should be set only by the admin, and we have still a lot of questions on how to use them between users see https://github.com/LibrePhotos/librephotos/issues/492

polaroidkidd commented on src/components/modals/ModalUserEdit.tsx:
```diff
@@ -0,0 +1,363 @@
+        </Grid.Col>
+      </Grid>
+      <Title order={6}>{t("modalscandirectoryedit.explanation3")}</Title>
+      <div style={{ height: "150px", overflow: "auto" }}>
```
AFAIK we have no set way of styling our components yet. Personally, I'm not a huge fan of inline-styles. We could opt for styled-components, or css modules perhaps? @derneuere , do you have any input here

derneuere replied:
This is code from the legacy component. So this would be an optional fix from my point of view. I think we wanted to refactor this component anyway, see https://github.com/LibrePhotos/librephotos/issues/555

I think we should stick to the paradigm Mantine uses, which would be styled-components, right?

polaroidkidd replied:
they have a number of different way to add styles. You can add them as [props](https://mantine.dev/styles/sx/), use the [createStyles](https://mantine.dev/styles/create-styles/) function or add styles via [classNames](https://mantine.dev/styles/styles-api/). Since we're not styling mantine components here, I think using the `createStyles` option would be the best fit.

polaroidkidd commented on src/components/modals/ModalUserEdit.tsx:
```diff
@@ -0,0 +1,363 @@
+          onChange={changedTreeData => setTreeData(changedTreeData)}
+          theme={FileExplorerTheme}
+          isVirtualized={false}
+          generateNodeProps={rowInfo => {
+            const nodeProps = {
+              onClick: event => nodeClicked(event, rowInfo),
+            };
+            if (selectedNodeId === rowInfo.node.id) {
+              // @ts-ignore
+              nodeProps.className = "selected-node";
+            }
+            return nodeProps;
+          }}
```
Again, these functions are redefined on every render. I'm not sure that it's the best approach.

derneuere replied:
This is code from the legacy component. So this would be an optional fix from my point of view.

polaroidkidd commented on src/components/modals/ModalUserEdit.tsx:
```diff
@@ -0,0 +1,363 @@
+export function ModalUserEdit(props: Props) {
+  const { isOpen, updateAndScan, selectedNodeId, onRequestClose, userList, createNew, firstTimeSetup } = props;
+  var userToEdit = props.userToEdit;
+  const [newScanDirectory, setNewScanDirectory] = useState("");
+  const [treeData, setTreeData] = useState([]);
+  const [userName, setUserName] = useState("");
+  const [userNameError, setUserNameError] = useState("");
+  const [userEmail, setUserEmail] = useState("");
+  const [userEmailError, setUserEmailError] = useState("");
+  const [userFirst, setUserFirst] = useState("");
+  const [userLast, setUserLast] = useState("");
+  const [userPassword, setUserPassword] = useState("");
+  const [newPasswordIsValid, setNewPasswordIsValid] = useState(true);
+
+  const [isAdmin, setIsAdmin] = useState(false);
+  const [scanDirectoryPlaceholder, setScanDirectoryPlaceholder] = useState("");
+  const dispatch = useAppDispatch();
+  const auth = useAppSelector(state => state.auth);
+  const { directoryTree } = useAppSelector(state => state.util);
+  const inputRef = useRef<HTMLInputElement>(null);
+  const { t } = useTranslation();
+  const [pathDoesNotExist, setPathDoesNotExist] = useState(false);
+  const [pathError, setPathError] = useState("");
+  const [closing, setClosing] = useState(false);
```
All the separate usages of `useState` will cause a re-render every time it changes. I think there's two options here.

1. Split up the component into smaller sub-components (one from each input field). 
2. Use the `useReducer` function here.

Personally, I'm in favor of `1` because it breaks this component up into smaller pieces, which are easier to digest when debugging. On the other hand, we'd have a hand-full of small components lying around which only get used once.

derneuere replied:
I am also in favor of splitting this up, but a lot of this is code from the legacy component. So this would be an optional fix from my point of view.

polaroidkidd commented on src/components/modals/ModalUserEdit.tsx:
```diff
@@ -0,0 +1,363 @@
+    } else {
+      const newData = replacePath(treeData, directoryTree[0]);
+
+      // @ts-ignore
```
why the `@ts-ignore` here?

derneuere replied:
This is code from the legacy component. So this would be an optional fix from my point of view.

polaroidkidd commented on src/components/modals/ModalUserEdit.tsx:
```diff
@@ -0,0 +1,363 @@
+
+export function ModalUserEdit(props: Props) {
+  const { isOpen, updateAndScan, selectedNodeId, onRequestClose, userList, createNew, firstTimeSetup } = props;
+  var userToEdit = props.userToEdit;
```
is it possible to use `let` here instead?

polaroidkidd commented on src/components/modals/ModalUserEdit.tsx:
```diff
@@ -0,0 +1,363 @@
+  const validateEmail = email => {
+    setUserEmailError("");
+
+    const emailRegex = // eslint-disable-next-line no-control-regex
+      /(?:[a-z0-9!#$%&'*+/=?^_`{|}~-]+(?:\.[a-z0-9!#$%&'*+/=?^_`{|}~-]+)*|"(?:[\x01-\x08\x0b\x0c\x0e-\x1f\x21\x23-\x5b\x5d-\x7f]|\\[\x01-\x09\x0b\x0c\x0e-\x7f])*")@(?:(?:[a-z0-9](?:[a-z0-9-]*[a-z0-9])?\.)+[a-z0-9](?:[a-z0-9-]*[a-z0-9])?|\[(?:(?:(2(5[0-5]|[0-4][0-9])|1[0-9][0-9]|[1-9]?[0-9]))\.){3}(?:(2(5[0-5]|[0-4][0-9])|1[0-9][0-9]|[1-9]?[0-9])|[a-z0-9-]*[a-z0-9]:(?:[\x01-\x08\x0b\x0c\x0e-\x1f\x ...
```
we should move this into a separate utils file

polaroidkidd commented on src/components/modals/ModalUserEdit.tsx:
```diff
@@ -0,0 +1,363 @@
+  firstTimeSetup?: boolean;
+};
+
+export function ModalUserEdit(props: Props) {
```
More of a general question, ss there a specific reason why we're not using `@mantine/form` here?

derneuere replied:
Would be nice, but most of the state is from the legacy component. So this would be an optional fix from my point of view.

polaroidkidd commented on src/layouts/settings/AdminPage.tsx:
```diff
@@ -60,9 +56,73 @@ export const AdminPage = () => {
+      </td>
+      <td>{user.username}</td>
+      <td>{user.scan_directory ? user.scan_directory : t("adminarea.notset")}</td>
+      {matches && <td>{user.confidence ? user.confidence : t("adminarea.notset")}</td>}
+      {matches && <td>{user.photo_count}</td>}
+      {matches && <td>{moment(user.date_joined).fromNow()}</td>}
```
you can wrap these into a single

```tsx
{ matches && 
<>
<td>{user.confidence ? user.confidence : t("adminarea.notset")}</td>
<td>{user.photo_count}</td>
<td>{moment(user.date_joined).fromNow()}</td>
</>
}
```

polaroidkidd commented on src/layouts/settings/SiteSettings.tsx:
```diff
@@ -28,12 +32,12 @@ export const SiteSettings = () => {
         label={t("sitesettings.header")}
         onChange={() => dispatch(setSiteSettings({ allow_registration: !siteSettings.allow_registration }))}
-        checked={siteSettings.allow_registration}
+        checked={allowRegistration}
```
why do we no longer listen to the site settings here?

derneuere replied:
I think the issue was, that it didn't update in time. It resulted in values going back to the old value and then to the new value when saving. Now it's in a state and does update correctly.

jakekapellen replied:
That's correct :-)

polaroidkidd commented on src/layouts/settings/Settings.tsx:
```diff
@@ -173,7 +173,7 @@ export const Settings = () => {
         radius="md"
       >
         <Text size="sm" style={{ marginBottom: 10 }} weight={500}>
```
can we use the `const { t } = useTranslation()` hook here?

derneuere replied:
This is code from the legacy component. So this would be an optional fix from my point of view.

jakekapellen replied:
Fixed on my machine. I'll commit to my branch today.

polaroidkidd commented on src/layouts/settings/Profile.tsx:
```diff
@@ -47,6 +48,17 @@ export function Profile() {
+
   // open update dialog, when user was edited
   useEffect(() => {
     if (JSON.stringify(userSelfDetailsRedux) !== JSON.stringify(userSelfDetails)) {
```
we have the lodash dependency. you can use [isEqual](https://lodash.com/docs/4.17.15#isEqual) here

jakekapellen replied:
Legacy code, but I'll update it anyway

polaroidkidd commented on src/components/settings/PasswordEntry.tsx:
```diff
@@ -0,0 +1,101 @@
+              style={{ marginLeft: "5px" }}
+              onClick={() => {
+                setEditPasswordMode(!editPasswordMode);
+              }}
+            >
+              {editPasswordMode ? <LockOpen size={16} /> : <Lock size={16} />}
+            </ActionIcon>
+          </Text>
+        )}
+      </Title>
+
+      <PasswordInput
+        icon={<Lock />}
+        placeholder={t("login.passwordplaceholder")}
+        name="password"
+        disabled={!editPasswordMode && !createNew}
+        required={editPasswordMode}
+        value={newPassword}
+        error={newPasswordError}
+        onChange={event => {
+          setNewPassword(event.currentTarget.value);
+          validateAndUpdatePassword(event.currentTarget.value, newPasswordConfirm);
+        }}
+      />
+      <PasswordInput
+        icon={<Lock />}
+        placeholder={t("login.confirmpasswordplaceholder")}
+        name="passwordConfirm"
+        disabled={!editPasswordMode && !createNew}
+        required={editPasswordMode}
+        value={newPasswordConfirm}
+        error={confirmPasswordError}
+        onChange={event => {
+          setNewPasswordConfirm(event.currentTarget.value);
+          validateAndUpdatePassword(newPassword, event.currentTarget.value);
+        }}
+      />
+    </Stack>
+  );
+}
```
can we use the `@mantine/form` dependency here?

derneuere replied:
I think that would make sense. It also reduces a lot of the validation logic. You can check out the login component as an example: https://github.com/LibrePhotos/librephotos-frontend/blob/dev/src/layouts/login/LoginPage.tsx

polaroidkidd commented on src/store/user/user.zod.ts:
```diff
@@ -46,9 +47,13 @@ export const ManageUser = z.object({
   photo_count: z.number(),
   save_metadata_to_disk: z.string(),
-  scan_directory: z.string(),
+  scan_directory: z.string().optional().nullable(),
```
is it possible to use zod's [nullish](https://github.com/colinhacks/zod#nullish) here?

derneuere commented:
I think some of the issues poloroidkidd found were already present in the old component. I will look over them tomorrow to figure out which are actually new issues.

jakekapellen committed:
Remove settings.json

derneuere commented:
Alright, left a couple of comments on issues that were already present in the legacy component. Would be nice if you have the time to fix them, but they are not necessary for merging :+1:

jakekapellen committed:
Significant rework, move user stuff to RTK

jakekapellen committed:
Merge branch 'dev' into dev

jakekapellen committed:
switch to nullish

jakekapellen committed:
fix merge issue

jakekapellen commented:
Big update! I've made some significant changes to respond to comments and add more value to this overall pull request.

Fixes:
- Moved from openConfirmModal back to a normal modal to remove the dependency on Mantine 5.0
- Moved EditUserModal to use a form
-  Removed settings.json and updated .gitignore to reflect that
- More translation fixes
- Lots of other tweaks you both requested.

Enhancements:
- Moved fetchUserList, manageUser (post), signup (post), and deleteUser to RTK Query, cleaning up a lot of code
- Added dependencies between the above queries/mutations to auto-refetch the user list when we invalidate the cache
- Added email/duplicate username checking to the setup/signup pages

jakekapellen committed:
minor tweak

jakekapellen committed:
another minor tweak

jakekapellen committed:
Fix uppercase usernames

derneuere commented:
Awesome! Looks good to me :+1:
